flink-connectors/flink-connector-base/src/main/java/org/apache/flink/connector/base/source/reader/fetcher/SplitFetcher.java

```diff
@@ -82,6 +85,14 @@ public class SplitFetcher<E, SplitT extends SourceSplit> implements Runnable {
 
     private final Consumer<Collection<String>> splitFinishedHook;
 
+    /**
+     * A shutdown latch to help make sure the SplitReader is only closed after all the emitted
+     * records have been processed by the main reader thread. This is needed because in some cases,
+     * the records in the <tt>RecordsWithSplitIds</tt> may have not been processed when the split
+     * fetcher shuts down.
+     */
+    private final CountDownLatch recordsProcessedLatch;
```

Comment:
nit: I think this patch introduces one more state for SplitFetcher: closed but not fully shutdown. It looks like SplitFetcher now has a lot of states, like running, paused, closed and shutdown. Maybe we can clean up the code of SplitFetcher in the future to make it more maintainable.